lib/deep-mark-required/index.ts

```diff
@@ -0,0 +1,30 @@
+import { AreNonDistributiveEqual } from "../are-non-distributive-equal";
+import { HasParsablePath } from "../has-parsable-path";
+import { MarkRequired } from "../mark-required";
+import { Paths } from "../paths";
+import { Prettify } from "../prettify";
+import { UnionToTuple } from "../union-to-tuple";
+
+type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
+  ? {
+      [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
+        ? Prettify<RecursiveDeepMarkRequiredSinglePath<NonNullable<Type[Key]>, RestKeyPath>>
+        : Type[Key];
+    }
+  : KeyPath extends keyof Type
+  ? Prettify<MarkRequired<Type, KeyPath>>
+  : Type;
+
+type RecursiveDeepMarkRequired<Accumulator, KeyPaths extends string[]> = KeyPaths extends [
+  infer KeyPath,
+  ...infer RestKeyPaths
+]
+  ? RestKeyPaths extends string[]
+    ? // Keep all the changes in `Accumulator`
+      RecursiveDeepMarkRequired<RecursiveDeepMarkRequiredSinglePath<Accumulator, KeyPath>, RestKeyPaths>
+    : never
+  : Accumulator;
```

Comment:
Upon further investigation, discovered that there's multiple versions of Prettier in `yarn.lock`.

<img width="889" alt="image" src="https://github.com/user-attachments/assets/67fa99ec-571a-4011-ae3d-e57e52d7ef00">

<br />
<br />

Since both `changeset` and our project allow upgrading the minor version, simply running `yarn upgrade prettier` would fix the issue.

<img width="919" alt="image" src="https://github.com/user-attachments/assets/eff76f4a-9cba-4f73-b7fb-366275ab70e4">

Reply:
This syntax appeared in TypeScript 4.7, but we still have to support 4.5 and 4.6 versions

```ts
A extends [B, infer C extends D]
              ^^^^^^^^^^^^^^^^^
```

I documented it in https://github.com/ts-essentials/ts-essentials/pull/408/commits/98d84ef38b4d5429278dfd5c213f037b27a97f85

Reply:
@som-sm I fixed yarn.lock in https://github.com/ts-essentials/ts-essentials/commit/b34cb4913001699b4b2868164ac20966c00e6272, thanks for finding such a specific issue

Reply:
@Beraliv Oh yeah, I missed that this syntax would fail in 4.5 and 4.6.

Reply:
No problems, it's a headache to remember about multiple TypeScript versions 😅 but it is a strategy of this utility library I'd like to keep